docs/proposals/llm-hyperparameter-optimization-api.md

```diff
@@ -0,0 +1,248 @@
+  * [Non-Goals](#non-goals)
+  * [Design for API](#design-for-api)
+    + [Example](#example)
+  * [Implementation](#implementation)
+
+## Links
+
+- [katib/issues#2291 (Tuning API in Katib for LLMs)](https://github.com/kubeflow/katib/issues/2291)
+
+## Motivation
+
+The rapid advancements and growing popularity of Large Language Models (LLMs) have driven an increased need for effective LLMOps in Kubernetes environments. To address this, we developed a [train API](https://www.kubeflow.org/docs/components/training/user-guides/fine-tuning/) within the Training Python SDK, simplifying the process of fine-tuning LLMs using distributed PyTorchJob workers. However, hyperparameter optimization remains a crucial yet labor-intensive task for enhancing model performance. Automating this tuning process through a dedicated API will facilitate efficient and scalable exploration of hyperparameters, ultimately improving model performance and reducing manual effort.
+
+## Goals
+
+Our goal is to develop a high-level API for tuning hyperparameters of LLMs that simplifies the process of hyperparameter optimization in Kubernetes. This API will seamlessly integrate with external platforms like HuggingFace and S3 for importing pretrained models and datasets. By specifying parameters for the training objective, trial configurations, and PyTorch worker configurations, the API will automate the creation of experiments and execution of trials. This abstraction of Kubernetes infrastructure complexities will enable data scientists to optimize hyperparameters efficiently and effectively.
+
+## Non-Goals
+
+1. Incorporate early stopping strategy into the API to optimize training efficiency.
+2. Expand support for distributed training in frameworks beyond PyTorch by leveraging their distributed training capabilities.
+3. Support adding custom providers through configmap or CRD approach to enhance flexibility.
+4. Enable users to deploy tuned models for inference within their applications or seamlessly integrate them into existing NLP pipelines for specialized tasks.
+
+## Design for API
+
+![Design for API](images/design_api.jpg)
+
+```python
+import kubeflow.katib as katib
+from kubeflow.katib import KatibClient
+
+class KatibClient(object):
+	
+	def tune(
+		self,
+		name: str, 
+		namespace: Optional[str] = None,
+		model_provider_parameters: Optional[kubeflow.storage_initializer.hugging_face.HuggingFaceModelParams] = None,
+		dataset_provider_parameters: Optional[kubeflow.storage_initializer.hugging_face.HuggingFaceDatasetParams, kubeflow.storage_initializer.s3.S3DatasetParams] = None,
+		trainer_parameters: Union[kubeflow.storage_initializer.HuggingFaceTrainerParams, Dict[str, Any]] = None,
+		storage_config: Dict[str, Optional[Union[str, List[str]]]] = {
+            "size": constants.PVC_DEFAULT_SIZE,
+            "storage_class": None,
+            "access_modes": constants.PVC_DEFAULT_ACCESS_MODES,
+        },
+		objective: Optional[Callable] = None,
```

Comment:
In addition to using the Storage Initializer to download models and datasets from HuggingFace or S3, I added `objective` and `base_image` parameters to allow users to define their own objective function. This approach enables us to modify the existing `tune` API directly instead of developing another one. What do you think? @andreyvelich @johnugeorge

Reply:
Yes, we should validate that user sets one of `objective` or `trainer_parameters` argument in the `tune` API. Other idea might be to give user ability to set `HuggingFaceTrainerParams` directly in objective argument. E.g.:
```python
objective = HuggingFaceTrainerParams(
		training_parameters = transformers.TrainingArguments(
			output_dir = "test_trainer",
			save_strategy = "no",
			eval_strategy = "epoch",
			disable_tqdm = True,
			log_level = "info",
			learning_rate = katib.search.double(min=1e-05, max=5e-05),
			per_device_train_batch_size = katib.search.categorical([8, 16, 32]),
			num_train_epochs = katib.search.int(min=1, max=10),
			weight_decay = katib.search.double(min=0.0, max=1.0),
		),
		# Set LoRA config to reduce number of trainable model parameters.
		lora_config = LoraConfig(
			r = katib.search.int(min=8, max=32),
			lora_alpha = 8,
			lora_dropout = 0.1,
			bias = "none",
		),
	),	
```

I understand that this will be in-consistent with `train` API in training client, but it will help us avoid unnecessary validation for one of 
WDYT @helenxie-bit @johnugeorge ?

Reply:
From my perspective, it's better to define `objective` and `trainer_parameters` separately. In this way, we can keep the `trainer_parameters` argument (instead of setting `HuggingFaceTrainerParams` directly in the `objective` argument, which would render `trainer_parameters` useless), and this approach is also clearer for users. Users will have two options:

1) Use external models and datasets: Users need to set `model_provider_parameters`, `dataset_provider_parameters`, `storage_config`, and `trainer_parameters`.
2) Define a custom objective function: Users need to set `objective` and `trainer_parameters`.

We can add validation for this logic, and it should not be difficult.

Reply:
@andreyvelich @johnugeorge According to today's discussion, I will keep these parameters as they are for now. We can make adjustments in the next version.